nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/CuratorHeartbeatMonitor.java

```diff
@@ -0,0 +1,376 @@
+    private static final Logger logger = LoggerFactory.getLogger(CuratorHeartbeatMonitor.class);
+    private static final Unmarshaller unmarshaller;
+
+    private final ClusterCoordinator clusterCoordinator;
+    private final ZooKeeperClientConfig zkClientConfig;
+    private final String heartbeatPath;
+    private final int heartbeatIntervalMillis;
+
+    private volatile CuratorFramework curatorClient;
+    private volatile ScheduledFuture<?> future;
+    private volatile Map<NodeIdentifier, NodeHeartbeat> latestHeartbeatMessages;
+    private volatile long latestHeartbeatTime;
+
+    private final FlowEngine flowEngine = new FlowEngine(1, "Heartbeat Monitor", true);
+
+    static {
+        try {
+            final JAXBContext jaxbContext = JAXBContext.newInstance(HeartbeatMessage.class);
+            unmarshaller = jaxbContext.createUnmarshaller();
+        } catch (final Exception e) {
+            throw new RuntimeException("Failed to create an Unmarshaller for unmarshalling Heartbeat Messages", e);
+        }
+    }
+
+    public CuratorHeartbeatMonitor(final ClusterCoordinator clusterCoordinator, final Properties properties) {
+        this.clusterCoordinator = clusterCoordinator;
+        this.zkClientConfig = ZooKeeperClientConfig.createConfig(properties);
+        this.heartbeatPath = zkClientConfig.resolvePath("cluster/heartbeats");
+
+        final String heartbeatInterval = properties.getProperty(NiFiProperties.CLUSTER_PROTOCOL_HEARTBEAT_INTERVAL,
+            NiFiProperties.DEFAULT_CLUSTER_PROTOCOL_HEARTBEAT_INTERVAL);
+
+        this.heartbeatIntervalMillis = (int) FormatUtils.getTimeDuration(heartbeatInterval, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void start() {
+        final RetryPolicy retryPolicy = new RetryForever(5000);
+        curatorClient = CuratorFrameworkFactory.newClient(zkClientConfig.getConnectString(),
+            zkClientConfig.getSessionTimeoutMillis(), zkClientConfig.getConnectionTimeoutMillis(), retryPolicy);
+        curatorClient.start();
+
+        this.future = flowEngine.scheduleWithFixedDelay(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    monitorHeartbeats();
+                } catch (final Exception e) {
+                    clusterCoordinator.reportEvent(null, Severity.ERROR, "Failed to process heartbeats from nodes due to " + e.toString());
+                    logger.error("Failed to process heartbeats", e);
+                }
+            }
+        }, heartbeatIntervalMillis, heartbeatIntervalMillis, TimeUnit.MILLISECONDS);
+    }
+
+    private CuratorFramework getClient() {
+        return curatorClient;
+    }
+
+    @Override
+    public void stop() {
+        final CuratorFramework client = getClient();
+        if (client != null) {
+            client.close();
+        }
+
+        if (future != null) {
+            future.cancel(true);
+        }
+    }
```

Comment:
re: FlowEngine - threads are daemon so they won't prevent shut down. However, it should be shutdown anyway. Will address.
client.close() also shouldn't ever throw an Exception, but it is probably worth canceling the future in a finally block.